src/java.base/share/classes/sun/security/ssl/CertificateMessage.java

```diff
@@ -1043,6 +1044,7 @@ private static SSLPossession choosePossession(
             }
 
             Collection<String> checkedKeyTypes = new HashSet<>();
+            LinkedHashSet<String> allAuths = new LinkedHashSet<>();
```

Comment:
Good suggestion.

One more thing: `checkedKeyTypes` only looks at `ss.keyAlgorithm`. I know the other checks (`SignatureScheme.getPreferableAlgorithm` and `X509Authentication.valueOf`) also only look at `ss.keyAlgorithm`, but are we going to check for more (Ex: group name) later? In the meantime, I would suggest changing the parameter type of these methods from `SignatureScheme` to `String` so we know only `keyAlgorithm` is checked.

Reply:
Did you mean to change hc.peerRequestedCertSignSchemes to String?  This field would be used for algorithm constraints and more parameters are required there.

Reply:
No, I meant changing the type of the `certScheme` argument in
```
    static SignatureScheme getPreferableAlgorithm(
            AlgorithmConstraints constraints,
            List<SignatureScheme> schemes,
            SignatureScheme certScheme,
            ProtocolVersion version) {
```
to `String`. Since we are only putting `ss.keyAlgorithm` value into the checked type list, we need to make sure we only check on about `keyAlgorithm` now. My concern is that suppose one day we decide to check on `ss.namedGroup` as well, then we will remember to update this method and the checked list at the same time.

Reply:
I don't think so.  In logic, we should check the SignatureScheme here rather than the String key algorithm.  If we need an update in the future, we may want to update the following-on methods, like key manager APIs.

Reply:
I agree it's not a problem now, and once we think it's a problem it will be a big one. I'll keep the current methods as is.